src/HDF5.jl

```diff
@@ -1370,12 +1378,56 @@ Base.getindex(parent::Union{File,Group}, r::Reference) = _deref(parent, r)
 Base.getindex(parent::Dataset, r::Reference) = _deref(parent, r) # defined separately to resolve ambiguity
 
 # convert special types to native julia types
-normalize_types(x) = x
-normalize_types(x::NamedTuple{T}) where {T} = NamedTuple{T}(map(normalize_types, values(x)))
-normalize_types(x::Cstring) = unsafe_string(x)
-normalize_types(x::FixedString) = unpad(String(collect(x.data)), pad(x))
-normalize_types(x::FixedArray) = normalize_types.(reshape(collect(x.data), size(x)...))
-normalize_types(x::VariableArray) = normalize_types.(copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false)))
+function normalize_types(::Type{T}, buf::AbstractMatrix{UInt8}) where {T}
+    return [_normalize_types(T, view(buf, :, ind)) for ind in axes(buf, 2)]
+end
+
+# high-level description which should always work --- here, the buffer contains the bytes
+# for exactly 1 element of an object of type T, so reinterpret the `UInt8` vector as a
+# length-1 array of type `T` and extract the (only) element.
+function _typed_load(::Type{T}, buf::AbstractVector{UInt8}) where {T}
+    return @inbounds reinterpret(T, buf)[1]
+end
+# fast-path for common concrete types with simple layout (which should be nearly all
+# cases)
+function _typed_load(::Type{T}, buf::V) where {T, V <: Union{Vector{UInt8}, Base.FastContiguousSubArray{UInt8,1}}}
+    dest = Ref{T}()
+    GC.@preserve buf begin
+        srcptr  = pointer(buf)
+        destptr = unsafe_convert(Ptr{Cvoid}, dest)
+        Base._memcpy!(destptr, srcptr, sizeof(T))
+    end
```

Comment:
```suggestion
    GC.@preserve buf Base._memcpy!(unsafe_convert(Ptr{Cvoid}, dest), pointer(buf), sizeof(T))
```
Sorry about this, but feel free to change this back